Dynamic RuntimeClass implementation

tallclair opened the pull request:
**What this PR does / why we need it**:

Implement RuntimeClass using the dynamic client to break the dependency on https://github.com/kubernetes/kubernetes/pull/67791

Once (if) https://github.com/kubernetes/kubernetes/pull/67791 merges, I will migrate to the typed client.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
For https://github.com/kubernetes/features/issues/585

**Release note**:
Covered by #67737
```release-note
NONE
```

/sig node
/kind feature
/priority important-soon
/milestone v1.12

tallclair committed:
Implement RuntimeClass support for the Kubelet & CRI

caesarxuchao commented on pkg/kubelet/runtimeclass/runtime_class_controller.go:
```diff
@@ -0,0 +1,89 @@
+func NewController(client dynamic.Interface) *Controller {
+	rc := client.Resource(runtimeClassGVR)
+	lw := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
```
Why is the wrapper needed?

caesarxuchao commented on pkg/kubelet/runtimeclass/runtimeclass_manager.go:
```diff
@@ -0,0 +1,89 @@
+/*
```
Is this file redundant?

tallclair reviewed (commented):
Thanks Chao.

tallclair commented:
Ready for review.

/assign @yujuhong 
/cc @dchen1107 @caesarxuchao @Random-Liu

tallclair commented:
/cc @saad-ali @jsafrane 
If you find yourselves blocked by getting `k8s.io/csi-api` checked in, this is an alternative approach (see pkg/kubelet/runtimeclass/runtimeclass_manager.go)

tallclair commented:
Fixed golint & rebased.

tallclair commented:
/assign @Random-Liu

tallclair commented:
Shoot, it looks like I pushed an old version by mistake. Will ping once resolved.

tallclair commented:
Actually it was just a bad rebase. Fixed. PTAL

Random-Liu commented on cmd/kubelet/app/server.go:
```diff
@@ -581,6 +584,10 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
 		}
+		dynamicKubeClient, err = dynamic.NewForConfig(clientConfig)
+		if err != nil {
+			glog.Errorf("Failed to initialize dynamic KubeClient: %v", err)
```
It seems that we use `Warningf` for the other clients.

tallclair replied:
Done.

Random-Liu commented on cmd/kubelet/app/server.go:
```diff
@@ -581,6 +584,10 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
 		}
+		dynamicKubeClient, err = dynamic.NewForConfig(clientConfig)
+		if err != nil {
+			glog.Errorf("Failed to initialize dynamic KubeClient: %v", err)
```
Other clients only generate `Warningf`.

Random-Liu reviewed (commented):
LGTM overall.

Based on https://github.com/kubernetes/kubernetes/pull/67009, dockershim doesn't support kata now even with runtime handler.

We may want to get the ip cache thing in as well, or not support runtime handler in dockershim at all and return error.

Also /cc @miaoyq

miaoyq commented:
Thanks @Random-Liu 
The motivation of this feature is to provide a way to select between different oci runtimes.
So `dockershim`, as the default CRI implementation for k8s, should support this feature IMO.

I also hope that `dockershim` can support this feature. :-)

tallclair commented:
On further consideration, I think the dockershim support should be split out to a separate PR. I reverted the changes and added an explicit error for the non-empty handler.

The problem with the original approach is that the RuntimeHandler is treated as a binary to call (without arguments). However, there is no way to specify the path of the binary, so it must be in the system path already. This may be OK, but still deserves its own discussion.

dchen1107 commented:
/lgtm

Please rebase it. We are good to go.

tallclair commented:
squashed, rebased, and fixed dockershim error message

dchen1107 commented:
/lgtm
/approve

lukaszo commented:
Not sure where to ask this question so asking here. Why RuntimeClass is passed to CRI and not handled by kubelet only?

tallclair commented:
Bumping to priority to enable merging. This was in the queue prior to freeze.
/priority critical-urgent

tallclair commented:
@lukaszo The runtime handler selects the configuration the CRI implementation should use. There is still only a single CRI server that the kubelet talks to.  As we add more features to RuntimeClass (e.g. pod overhead, scheduling & discovery, supported features, etc.) the Kubelet will do more of the work.
